lib/Dialect/FIRRTL/Transforms/LowerDPI.cpp

```diff
@@ -0,0 +1,158 @@
+//===- LowerDPI.cpp - Lower to DPI to Sim dialects ------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the LowerDPI pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLDialect.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/FIRRTL/Namespace.h"
+#include "circt/Dialect/Sim/SimOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Threading.h"
+#include "mlir/Support/LogicalResult.h"
+#include "llvm/ADT/MapVector.h"
+
+using namespace mlir;
+using namespace llvm;
+using namespace circt;
+using namespace circt::firrtl;
+
+struct LowerDPIPass : public LowerDPIBase<LowerDPIPass> {
+  void runOnOperation() override;
+};
+
+void LowerDPIPass::runOnOperation() {
+  auto circuitOp = getOperation();
+
+  CircuitNamespace nameSpace(circuitOp);
+  MapVector<StringAttr, SmallVector<DPICallIntrinsicOp>> funcNameToCallSites;
+  {
+    // A helper struct to collect DPI calls in the circuit.
+    struct DpiCallCollections {
+      FModuleOp module;
+      SmallVector<DPICallIntrinsicOp> dpiOps;
+    };
+    SmallVector<DpiCallCollections> results;
```

Comment:
Consider a `reserve`.

Reply:
Since we cannot know the actual number of FModuleOp beforehand I reserved 64.

Reply:
Probably 64 won't fit on the stack / small size, so maybe make the SmallVector 0-length (or 64?) or whatever seems good.  Not blocking, just a stray thought since tab still open :).